aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,427 @@
+        self._querybuilder.add_projection(self._last_tag, self._entity_to_identifier)
+
+        if self._track_edges:
+            # I am keeping track of the edges I traverse.
+            # This requires additional projections:
+            projections = defaultdict(list)
+
+            # Need the edge_keys, which is a list of tuples (of length 2) that stores the information
+            # what I need to project as well, as in (tag, projection)
+            self._edge_keys = []
+            self._edge_label = queryhelp['path'][-1]['edge_tag']
+
+            # Need to get the edge_set: This is given by entity1_entity2. Here, the results needs to
+            # be sorted somehow in order to ensure that the same key is used when entity_from and
+            # entity_to are exchanged.
+            edge_set = operational_set.dict['{}_{}'.format(*sorted((self._entity_from, self._entity_to)))]
+
+            # Looping over the edge identifiers to figure out what I need to project and in which
+            # order. The order is important! The rules:
+            # r1 = Rule(QueryBuilder().append(Group).append(Node, with_group=Group) and
+            # r2 = Rule(QueryBuilder().append(Node).append(Group, with_node=Node)
+            # need still to save their results in the same order (i.e. group_id, node_id).
+            # Therefore, I am sorting the edge_keys according to edge_identifiers
+            # specified in the edge_set
+            for tag, projection in edge_set.edge_identifiers:
+                if tag == 'edge':
+                    actual_tag = self._edge_label
+                elif tag == self._entity_from:
+                    actual_tag = self._first_tag
+                elif tag == self._entity_to:
+                    actual_tag = self._last_tag
+                else:
+                    # For now I can only specify edge_identifiers as 'edge', ie. project on the edge
+                    # itself, or by the entity_from, entity_to keyword, ie. groups or nodes.
+                    # One could think of other keywords...
+                    raise ValueError('This tag ({}) is not known'.format(tag))
+                self._edge_keys.append((actual_tag, projection))
+                projections[actual_tag].append(projection)
+            # Telling the QB about the additional projections:
+            for proj_tag, projectionlist in projections.items():
+                try:
+                    self._querybuilder.add_projection(proj_tag, projectionlist)
+                except InputValidationError:
+                    raise KeyError('The projection for the edge-identifier is invalid.\n')
+
+    def _load_results(self, target_set, operational_set):
+        """Single application of the rules to the operational set
+
+        :param target_set: where the results will be loaded
+        :param operational_set: where the results originate from (walkers)
+        """
+        # I check that I have primary keys
```

Comment:
This comment and the one below - generally all in this area of the code, should be revisited, as they mostly just say what the code also says in itself, i.e., they are unnecessary, duplicating the meaning of the code.

Reply:
I cleaned up a bit more of the comments here, and I think most of the methods are ok now. The only one that remains heavily commented is `_init_run` of `QueryRule`, but since this is the core of the recursive engine and the code is written in a very "condensed" way, I am hesitant to remove much of this yet. I would like to eventually streamline the whole procedure, but I don't think I can manage to that in the immediate future (i.e. this PR would be too delayed if I got into that).